### ISSUE TITLE
[chip-test] chip_sw_otbn_mem_scramble

### ISSUE THREAD

**Issue by GregAC** [lowRISC/opentitan#20119]
### Test point name

[chip_sw_otbn_mem_scramble](https://github.com/lowRISC/opentitan/blob/d1676c071ab51d2f4921beaa798fe7148b890c90/hw/top_earlgrey/data/ip/chip_otbn_testplan.hjson#L101-L129)

### Host side component

None

### OpenTitanTool infrastructure implemented

None

### Silicon Validation (SiVal)

Yes

### Emulation Targets

- [ ] None
- [X] CW310
- [ ] Hyperdebug + CW310

### Contact person

@GregAC

### Checklist

Please fill out this checklist as items are completed. Link to PRs and issues as appropriate.
- [ ] Check if existing test covers most or all of this testpoint (if so, either extend said test to cover all points, or skip the next 3 checkboxes)
- [ ] Device-side (C) component developed
- [ ] Bazel build rules developed
- [ ] Host-side component developed
- [ ] HJSON test plan updated with test name (so it shows up in the dashboard)
- [ ] Test added to dvsim nightly regression (and passing at time of checking)
- [ ] For SiVal test cases, test is running relevant FPGA or silicon regression

> estimate 0.5

**Comment by HU90m** [comment]
The existing `//sw/device/tests:otbn_mem_scramble_test` hasn't run in CI for as far back as the history goes.

After scrambling, ECC errors when reading values from IMEM should trigger interrupts but they don't appear to be:

```
I00051 otbn_mem_scramble_test.c:180] Mismatch without integrity error: Entry 49, address 0x29f, data 0x6984800c
E00052 otbn_mem_scramble_test.c:244] CHECK-fail: Expecting at least 48 IMEM integrity errors, got 0
I00053 status.c:34] FAIL!
```

**Comment by luismarques** [comment]
This test works everywhere except the FPGA, and *very occasionally* also fails on dvsim.

I found a couple of dvsim VCS seeds that fail:

- 105706649916084870360541638213197649019961057817534949222100493257430194194456
- 84178660973699573356153380528591368847312397780491565840024634280044710784108

dvsim reproducer:

```
./util/dvsim/dvsim.py hw/top_earlgrey/dv/chip_sim_cfg.hjson -i chip_sw_otbn_mem_scramble --tool vcs --reseed 1 --seeds 105706649916084870360541638213197649019961057817534949222100493257430194194456
```

The test itself seems correct. The NMI that should fire to make the test work seems not to fire on the FPGA (and, presumably, dvsim, I haven't checked the waveforms yet), but I don't know why.

**Comment by luismarques** [comment]
@vogelpi After figuring out some impediments that were preventing running tests with older commits, I managed to go back to your rework of the `otbn_mem_scramble_test` in commit `e0997d1c32` and it seems like this test also didn't pass on the FPGA at that time. So it seems like this test has *not* been broken by subsequent changes, but instead this is an issue related to how the FPGA behaves differently in ways that matter for your test. I also tried some of your previous commits around the same date and the results are the same. Here are the steps to reproduce, as they are non-obvious in some ways:

```
git checkout e0997d1c32
git cherry-pick --no-commit 609d0f5fe2 # you'll probably need this to program the CW310 FPGA with the current SAM3X firmware
./bazelisk.sh test //sw/device/tests:otbn_mem_scramble_test_fpga_cw310 --test_output=streamed --define bitstream=vivado
```

You mentioned some possibly optimisations for the FPGA that might be related to this. Can you or any of the HW folks investigate this?

Also, I'm curious. Do you remember if you ran this test on the FPGA at all back then (i.e. did it ever pass on the FPGA), or were you relying on some other test target (verilator?)? Based on @HU90m's comment above, it seems like this never ran in CI (i.e. the FPGA), but I'm not sure what you were using for this test's development.

**Comment by vogelpi** [comment]
Thanks @luismarques for your investigation. The questionable PR is https://github.com/lowRISC/opentitan/pull/14530 and it seems I mentioned that this needs to be tested on FPGA and with VCS and later that this is successfully working with Verilator and VCS (without mentioning FPGA). To honest, I don't remember what my finding on the FPGA were.

I can certainly help investigating here. What is the exact failure you're seeing? Is it the one mentioned by @HU90m above or something else? As for the FPGA specific changes in OTBN: On FPGA, we 1) only advance the PRNG (URND) when actually reading from it as opposed to advancing it in every clock cycle and 2) we skip reseeding operations of the same PRNG at program start.

The first change really shouldn't impact the test, the second change will cause the test to start / run faster. Maybe that's related to the issue here?

**Comment by luismarques** [comment]
> What is the exact failure you're seeing?

The interrupts don't trigger on the FPGA when there is an integrity error. I can share more details tomorrow if needed.

**Comment by luismarques** [comment]
> > What is the exact failure you're seeing?
> 
> The interrupts don't trigger on the FPGA when there is an integrity error. That causes the failure that @HU90m reported.

@vogelpi I think this issue of the interrupts not being generated on the FPGA when there is an integrity error also affects the test `chip_sw_sram_scrambled_access`. Is there something else I can help you with to make progress on this issue? As far as I understand, at this point it needs to be fixed on the HW (FPGA) side, which is a bit out of my area of expertise :)

**Comment by luismarques** [comment]
> [@vogelpi](https://github.com/vogelpi) I think this issue of the interrupts not being generated on the FPGA when there is an integrity error also affects the test `chip_sw_sram_scrambled_access`. Is there something else I can help you with to make progress on this issue? As far as I understand, at this point it needs to be fixed on the HW (FPGA) side, which is a bit out of my area of expertise :)

Ping @vogelpi who should look at that? The test itself is complete, so perhaps a separate issue should be opened for that and this one should be closed.

**Comment by vogelpi** [comment]
Sorry @luismarques , I completely forgot about this. I have many other things to take care of now but @nasahlpa offered to look into this.

**Comment by nasahlpa** [comment]
I did quite some digging:

**TLDR:** The interrupt we are expecting here is only triggered when using the `SecureIbex` configuration, which we don't use for CW310/CW340:
https://github.com/lowRISC/opentitan/blob/64ceddcdf91a3cbd27beadf7084002ea79fb48b0/hw/top_earlgrey/rtl/autogen/chip_earlgrey_cw310.sv#L1084


By wiping IMEM/DMEM, data is wrongly de-scrambled, i.e., ECC bits and data do not match. However, ECC bits and data are not checked inside OTBN, they are transported over TLUL to the Ibex. In the `SecureIbex` configuration, the `MemECC` parameter is set. Then, in the LSU the data integrity check happens and the corresponding interrupt is triggered. However, as we do not use the SecureIbex configuration, MemECC is disabled meaning that the ECC bits are not considered inside the LSU:
https://github.com/lowRISC/opentitan/blob/64ceddcdf91a3cbd27beadf7084002ea79fb48b0/hw/vendor/lowrisc_ibex/rtl/ibex_load_store_unit.sv#L353-L355
Hence, no integrity error and therefore no interrupt is triggered.

@vogelpi am I right that we do not enable the `SecureIbex` config for CW310/CW340? If yes, I guess that is because of resource constraints? As this config is quite sensitive and has a large impact on the design, are we aware of this?

**Comment by vogelpi** [comment]
Thanks for your investigation @nasahlpa ! Yes,it's correct that we don't enable the lockstep countermeasure (`SecureIbex`) on the FPGA due to resource constraints. I wasn't aware that memory ECC depends on this.

I don't think it's feasible to enable `SecureIbex` on the CW310, but for the CW340, it could work. We should explore this option. 

Alternatively, we could try to selectively enable some of the countermeasures grouped under the `SecureIbex` parameter. I am thinking of memory and register file ECC, dummy instructions because these can have notable impact on software. But they many of them have limited area impact. The highest area impact has the lockstep and maybe register-file ECC. Maybe we should just a separate parameter for the lockstep?

**Comment by nasahlpa** [comment]
Thanks @vogelpi. I've created an issue for this (#25137) and already did some experiments.

@luismarques for now the test will not work on CW310/CW340. How do we handle this in the testplan?

**Comment by luismarques** [comment]
> Thanks [@vogelpi](https://github.com/vogelpi). I've created an issue for this ([#25137](https://github.com/lowRISC/opentitan/issues/25137)) and already did some experiments.
> 
> [@luismarques](https://github.com/luismarques) for now the test will not work on CW310/CW340. How do we handle this in the testplan?

The test is already marked as broken for the FPGA. Once it starts working for the CW340 I guess we'd tweak the exec environments to run there (but not on the CW310) and remove the broken tag. This is showing up on the SiVal spreadsheet as failing due to the way the tests are currently run (it also runs tests marked as broken), but @engdoreis is going to change that nightly runner to use the CI infrastructure so it should start respecting the broken tag.

**Comment by vogelpi** [comment]
I am closing this issue, this test has been enabled on FPGA with 0645c085ca49964430bbf822552899b262da8eae (#25537).
